File: server/scanner.js
// Core scanner: reads candles, computes indicators and a 0-100 score, and
// returns only early-pump signals. Works with any exchange adapter that
// implements { listActiveUsdtSymbols, getKlines, chartUrl }.

const DEFAULTS = {
  interval: "5m",
  limit: 50,
  minScore: 70,
  priceChangeLookback: 4, // last N closed candles for the "price change" metric
  volumeLookback: 20, // window for average volume comparison
  breakoutLookback: 20, // window for breakout highest-high
  overextendedPct: 10, // already-pumped cutoff (skip signals)
  concurrency: 8, // parallel klines fetches
};

function last(arr, n = 1) {
  return arr[arr.length - n];
}

export function analyze(symbol, candles, cfg = DEFAULTS) {
  if (!candles || candles.length < cfg.breakoutLookback + 2) return null;

  const latest = last(candles);
  const priorClose = last(candles, cfg.priceChangeLookback + 1)?.close ?? candles[0].close;
  const priceChangePct = ((latest.close - priorClose) / priorClose) * 100;

  // Volume spike: latest volume vs average of the previous `volumeLookback`
  const volWindow = candles.slice(-1 - cfg.volumeLookback, -1);
  const avgVolume = volWindow.reduce((s, c) => s + c.volume, 0) / volWindow.length;
  const volumeSpike = avgVolume > 0 ? latest.volume / avgVolume : 0;

  // Breakout: current close strictly above prior window high
  const breakoutWindow = candles.slice(-1 - cfg.breakoutLookback, -1);
  const priorHigh = Math.max(...breakoutWindow.map((c) => c.high));
  const breakout = latest.close > priorHigh;

  // Candle strength: how close the close is to the high of the candle
  const range = latest.high - latest.low;
  const body = latest.close - latest.open;
  const bullish = body > 0;
  const closeNearHigh = range > 0 ? (latest.close - latest.low) / range : 0; // 0..1

  // Momentum: consecutive green closes ending at the latest candle
  let greens = 0;
  for (let i = candles.length - 1; i >= 0; i--) {
    if (candles[i].close > candles[i].open) greens++;
    else break;
  }

  // Already-pumped guard: move over the breakout window is too large
  const lowWindow = Math.min(...breakoutWindow.map((c) => c.low));
  const windowMovePct = lowWindow > 0 ? ((latest.close - lowWindow) / lowWindow) * 100 : 0;
  const overextended = windowMovePct > cfg.overextendedPct * 2 || priceChangePct > cfg.overextendedPct;

  // Scoring (0-100):
  //   price change 0-30 | volume spike 0-30 | breakout 20 | candle strength 10 | momentum 10
  const priceScore = Math.max(0, Math.min(30, (priceChangePct / 5) * 30));
  const volumeScore = Math.max(0, Math.min(30, ((volumeSpike - 1) / 4) * 30));
  const breakoutScore = breakout ? 20 : 0;
  const candleScore = bullish ? Math.round(closeNearHigh * 10) : 0;
  const momentumScore = Math.min(10, Math.max(0, (greens - 1) * 5));

  const score = Math.round(priceScore + volumeScore + breakoutScore + candleScore + momentumScore);

  const earlyPump =
    !overextended &&
    priceChangePct >= 1.5 &&
    volumeSpike >= 2.5 &&
    breakout &&
    bullish;

  // Badges communicate the "why" in the UI
  const badges = [];
  if (volumeSpike >= 2.5) badges.push("Volume Explosion");
  if (breakout) badges.push("Breakout");
  if (greens >= 3) badges.push("Momentum");

  return {
    symbol,
    priceChange: Number(priceChangePct.toFixed(2)),
    volumeSpike: Number(volumeSpike.toFixed(2)),
    breakout,
    bullish,
    greenCandles: greens,
    closeNearHigh: Number(closeNearHigh.toFixed(2)),
    overextended,
    earlyPump,
    score,
    badges,
    price: latest.close,
    sparkline: candles.slice(-20).map((c) => c.close),
    timeframe: cfg.interval,
  };
}

async function runWithConcurrency(items, concurrency, worker) {
  const results = [];
  let i = 0;
  const runners = Array.from({ length: concurrency }, async () => {
    while (i < items.length) {
      const idx = i++;
      try {
        const out = await worker(items[idx], idx);
        if (out) results.push(out);
      } catch {
        // Per-symbol failures are swallowed: one bad klines response shouldn't
        // take down the whole scan.
      }
    }
  });
  await Promise.all(runners);
  return results;
}

export async function scan(exchange, opts = {}) {
  const cfg = { ...DEFAULTS, ...opts };
  const symbols = await exchange.listActiveUsdtSymbols({
    minQuoteVolume: cfg.minQuoteVolume,
    limit: cfg.symbolLimit,
  });

  const analyzed = await runWithConcurrency(symbols, cfg.concurrency, async (symbol) => {
    const candles = await exchange.getKlines(symbol, {
      interval: cfg.interval,
      limit: cfg.limit,
    });
    return analyze(symbol, candles, cfg);
  });

  const signals = analyzed
    .filter((a) => a && a.earlyPump && a.score >= cfg.minScore)
    .sort((a, b) => b.score - a.score)
    .map((a) => ({
      ...a,
      chartUrl: exchange.chartUrl(a.symbol),
      exchange: exchange.id,
    }));

  return {
    scannedAt: new Date().toISOString(),
    exchange: exchange.id,
    scanned: analyzed.length,
    total: symbols.length,
    signals,
  };
}

export { DEFAULTS };

File: deploy/scanner.py
"""Core scanner — exchange-agnostic pump detection.

Ported from server/scanner.js. Same inputs and formulas so that scores stay
comparable across the Node and Python versions.
"""
from __future__ import annotations

import asyncio
from dataclasses import dataclass
from datetime import datetime, timezone

import httpx


@dataclass
class ScanConfig:
    interval: str = "5m"
    limit: int = 50
    min_score: int = 70
    price_change_lookback: int = 4
    volume_lookback: int = 20
    breakout_lookback: int = 20
    overextended_pct: float = 10.0
    concurrency: int = 8
    min_quote_volume: float = 5_000_000
    symbol_limit: int = 120


def analyze(symbol: str, candles: list[dict], cfg: ScanConfig) -> dict | None:
    if not candles or len(candles) < cfg.breakout_lookback + 2:
        return None

    latest = candles[-1]
    prior_close_idx = max(0, len(candles) - (cfg.price_change_lookback + 1))
    prior_close = candles[prior_close_idx]["close"]
    price_change_pct = (
        (latest["close"] - prior_close) / prior_close * 100 if prior_close else 0.0
    )

    vol_window = candles[-1 - cfg.volume_lookback : -1]
    avg_vol = sum(c["volume"] for c in vol_window) / len(vol_window) if vol_window else 0
    volume_spike = latest["volume"] / avg_vol if avg_vol > 0 else 0.0

    breakout_window = candles[-1 - cfg.breakout_lookback : -1]
    prior_high = max(c["high"] for c in breakout_window)
    breakout = latest["close"] > prior_high

    candle_range = latest["high"] - latest["low"]
    body = latest["close"] - latest["open"]
    bullish = body > 0
    close_near_high = (
        (latest["close"] - latest["low"]) / candle_range if candle_range > 0 else 0.0
    )

    greens = 0
    for c in reversed(candles):
        if c["close"] > c["open"]:
            greens += 1
        else:
            break

    low_window = min(c["low"] for c in breakout_window)
    window_move_pct = (
        (latest["close"] - low_window) / low_window * 100 if low_window > 0 else 0.0
    )
    overextended = (
        window_move_pct > cfg.overextended_pct * 2
        or price_change_pct > cfg.overextended_pct
    )

    price_score = max(0, min(30, (price_change_pct / 5) * 30))
    volume_score = max(0, min(30, ((volume_spike - 1) / 4) * 30))
    breakout_score = 20 if breakout else 0
    candle_score = round(close_near_high * 10) if bullish else 0
    momentum_score = min(10, max(0, (greens - 1) * 5))
    score = round(price_score + volume_score + breakout_score + candle_score + momentum_score)

    early_pump = (
        not overextended
        and price_change_pct >= 1.5
        and volume_spike >= 2.5
        and breakout
        and bullish
    )

    badges: list[str] = []
    if volume_spike >= 2.5:
        badges.append("Volume Explosion")
    if breakout:
        badges.append("Breakout")
    if greens >= 3:
        badges.append("Momentum")

    return {
        "symbol": symbol,
        "priceChange": round(price_change_pct, 2),
        "volumeSpike": round(volume_spike, 2),
        "breakout": breakout,
        "bullish": bullish,
        "greenCandles": greens,
        "closeNearHigh": round(close_near_high, 2),
        "overextended": overextended,
        "earlyPump": early_pump,
        "score": score,
        "badges": badges,
        "price": latest["close"],
        "sparkline": [c["close"] for c in candles[-20:]],
        "timeframe": cfg.interval,
    }


async def _analyze_symbol(
    exchange, client: httpx.AsyncClient, symbol: str, cfg: ScanConfig
) -> dict | None:
    try:
        candles = await exchange.get_klines(
            client, symbol, interval=cfg.interval, limit=cfg.limit
        )
    except Exception:
        return None
    return analyze(symbol, candles, cfg)


async def scan(exchange, cfg: ScanConfig) -> dict:
    async with httpx.AsyncClient(http2=False) as client:
        symbols = await exchange.list_active_usdt_symbols(
            client, min_quote_volume=cfg.min_quote_volume, limit=cfg.symbol_limit
        )
        sem = asyncio.Semaphore(cfg.concurrency)

        async def _run(sym: str) -> dict | None:
            async with sem:
                return await _analyze_symbol(exchange, client, sym, cfg)

        analyzed = [a for a in await asyncio.gather(*(_run(s) for s in symbols)) if a]

    signals = sorted(
        (a for a in analyzed if a["earlyPump"] and a["score"] >= cfg.min_score),
        key=lambda a: a["score"],
        reverse=True,
    )
    for s in signals:
        s["chartUrl"] = exchange.chart_url(s["symbol"])
        s["exchange"] = exchange.id

    return {
        "scannedAt": datetime.now(timezone.utc).isoformat(),
        "exchange": exchange.id,
        "scanned": len(analyzed),
        "total": len(symbols),
        "signals": signals,
    }
